picard/acousticbrainz/__init__.py

```diff
@@ -0,0 +1,221 @@
+    tagger.webservice.get(
+        host=ACOUSTICBRAINZ_HOST,
+        port=ACOUSTICBRAINZ_PORT,
+        path="/%s/low-level" % recording_id,
+        handler=partial(run_extractor, input_path, extractor_callback),
+        priority=True,
+        important=False,
+        parse_response_type=None
+    )
+
+
+def ab_extractor_callback(tagger, file, result, error):
+    file.clear_pending()
+
+    ab_metadata_file, result, error = result.result() if isinstance(result, Future) else result
+    if "Writing results" in error and ab_metadata_file:
+        file.acousticbrainz_features_file = ab_metadata_file
+        tagger.log.debug("AcousticBrainz extracted features of recording %s: %s" %
+                  (file.metadata["musicbrainz_recordingid"], file.filename,))
+
+        # Submit results
+        ab_submit_features(tagger, file, ab_metadata_file)
+    elif 'Duplicate' in error:
+        tagger.log.debug("AcousticBrainz already has an entry for recording %s: %s" %
+                  (file.metadata["musicbrainz_recordingid"], file.filename,))
+        file.acousticbrainz_is_duplicate = True
+    else:
+        # Something went wrong
+        tagger.log.debug("AcousticBrainz extraction failed with error: %s" % error)
+        file.acousticbrainz_error = True
+        if result == 1:
+            tagger.log.warning("AcousticBrainz extraction failed: %s" % file.filename)
+        elif result == 2:
+            tagger.log.warning(
+                "AcousticBrainz extraction failed due to missing a MusicBrainz Recording ID: %s" % file.filename)
+        else:
+            tagger.log.warning("AcousticBrainz extraction failed due to an unknown error: %s" % file.filename)
+    file.update()
+
+
+def run_extractor(input_path, extractor_callback, response, reply, error, main_thread=False):
+    duplicate = False
+    # Check if AcousticBrainz server answered with the json file for the recording id
+    if not error:
+        # If it did, mark file as duplicate and skip extraction
+        try:
+            load_json(response)
+            duplicate = True
+        except json.JSONDecodeError:
+            pass
+
+    if duplicate:
+        # If an entry for the same recording ID already exists
+        # using the same extractor version exists, skip extraction
+        results = (None, 0, "Duplicate")
+        extractor_callback(results, None)
+    else:
+        if main_thread:
+            # Run extractor on main thread, used for testing
+            extractor_callback(extractor(input_path), None)
+        else:
+            # Run extractor on a different thread and call the callback when done
+            run_task(partial(extractor, input_path), extractor_callback)
+
+
+def extractor(input_path):
+    config = get_config()
+
+    # Create a temporary file with AcousticBrainz output
+    output_file = NamedTemporaryFile("w", suffix=".json")
+    output_file.close()  # close file to ensure other processes can write to it
```

Comment:
Does the temp file get removed somewhere? I think we should clean it up, or does anything speak against that?

Reply:
Technically this deletes the file right away, but I wanted a safe temporary filename to use. They won't be automatically deleted, just like the coverart files. =x

Reply:
Got it, so it actually gets really created later.

Cover art actually get deleted when Picard exits, see https://github.com/metabrainz/picard/blob/master/picard/coverart/image.py#L72 .

But probably we could remove these files directly after they have been used? The data is not so much, but potentially a lot of files will be created?

Reply:
I've seen this was there, but my temp folder always had tons of leftovers, so I assumed it didn't exactly work as intended. Testing it right now and it does work just fine, I was the one killing it with the IDE. Maybe an additional signal handler? =x